compose/ui/ui/src/webTest/kotlin/androidx/compose/ui/input/keydownEvent.kt

```
@@ -0,0 +1,65 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.ui.input
+
+import org.w3c.dom.events.KeyboardEvent
+import org.w3c.dom.events.KeyboardEventInit
+
+internal external interface KeyboardEventInitExtended : KeyboardEventInit {
+    var keyCode: Int?
+}
+
+private fun KeyboardEventInit.keyDownEvent() = KeyboardEvent("keydown", this)
+private fun KeyboardEventInit.withKeyCode() = (this as KeyboardEventInitExtended).apply {
+    keyCode = key!!.uppercase().first().code
+}
+
+internal fun keyDownEvent(
+    c: String,
+    code: String = "Key${c.uppercase()}",
+    ctrlKey: Boolean = false,
+    metaKey: Boolean = false,
+    altKey: Boolean = false,
+    shiftKey: Boolean = false,
+    cancelable: Boolean = true
+): KeyboardEvent =
+    KeyboardEventInit(
+        key = c,
+        code = code,
+        ctrlKey = ctrlKey,
+        metaKey = metaKey,
+        altKey = altKey,
+        shiftKey = shiftKey,
+        cancelable = cancelable
+    )
+        .withKeyCode()
+        .keyDownEvent()
+
+
+internal fun keyDownEvent(
+    c: Char = 'c',
+    code: String = "Key${c.uppercase()}",
+    ctrlKey: Boolean = false,
+    metaKey: Boolean = false,
+    altKey: Boolean = false,
+    shiftKey: Boolean = false,
+    cancelable: Boolean = true
+) = keyDownEvent("$c", code, ctrlKey, metaKey, altKey, shiftKey, cancelable)
+
+internal fun keyDownEventUnprevented(): KeyboardEvent =
+    KeyboardEventInit(ctrlKey = true, cancelable = true)
+        .keyDownEvent()
```

I guess the test using this function should fail in your PR too, due to missing key='Control'.

I fixed it in this PR: https://github.com/JetBrains/compose-multiplatform-core/pull/1303/files#diff-9bb816d7766408aff581d4414cb9c632ad5d611d355959011ff77fbe464db89dR219

---

Our PRs have a conflict. But not difficult to fix. We can merge your PR first.

---

yep, it was failing and I was exactly up to discuss it with you
so, let's discuss, if something is fixable on my side and wouldn't create additional burden on your side, let me just adopt it
also, we can merge you PR first - I have no problems with it and it's also fair - who's the first should no have a headache (even a little one) of resolving conflicts )